lucene/core/src/java/org/apache/lucene/store/MemorySegmentIndexInput.java

```diff
@@ -0,0 +1,525 @@
+import java.nio.ByteOrder;
+import java.util.Arrays;
+import java.util.Objects;
+
+import org.apache.lucene.util.IOUtils;
+
+import jdk.incubator.foreign.MemoryHandles;
+import jdk.incubator.foreign.MemorySegment;
+
+/**
+ * Base IndexInput implementation that uses an array of MemorySegments to represent a file.
+ *
+ * <p>For efficiency, this class requires that the segment size are a power-of-two
+ * (<code>chunkSizePower</code>).
+ */
+public abstract class MemorySegmentIndexInput extends IndexInput implements RandomAccessInput {
+  // We pass 1L as alignment, because currently Lucene file formats are heavy unaligned: :(
+  static final VarHandle VH_getByte = MemoryHandles.varHandle(byte.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getShort = MemoryHandles.varHandle(short.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getInt = MemoryHandles.varHandle(int.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  static final VarHandle VH_getLong = MemoryHandles.varHandle(long.class, 1L, ByteOrder.BIG_ENDIAN).withInvokeExactBehavior();
+  
+  final boolean isClone;
+  final long length;
+  final long chunkSizeMask;
+  final int chunkSizePower;
+  final MemorySegment[] segments;
+  
+  int curSegmentIndex = -1;
+  MemorySegment curSegment; // redundant for speed: segments[curSegmentIndex], also marker if closed!
+  long curPosition; // relative to curSegment, not globally
+
+  public static MemorySegmentIndexInput newInstance(
+      String resourceDescription,
+      MemorySegment[] segments,
+      long length,
+      int chunkSizePower) {
+    if (segments.length == 1) {
+      return new SingleSegmentImpl(resourceDescription, segments[0], length, chunkSizePower, false);
+    } else {
+      return new MultiSegmentImpl(resourceDescription, segments, 0, length, chunkSizePower, false);
+    }
+  }
+
+  private MemorySegmentIndexInput(
+      String resourceDescription,
+      MemorySegment[] segments,
+      long length,
+      int chunkSizePower,
+      boolean isClone) {
+    super(resourceDescription);
+    this.segments = segments;
+    this.length = length;
+    this.chunkSizePower = chunkSizePower;
+    this.chunkSizeMask = (1L << chunkSizePower) - 1L;
+    this.isClone = isClone;
+    this.curSegment = segments[0];
+  }
+  
+  void ensureOpen() {
+    if (curSegment == null) {
+      throw alreadyClosed();
+    }
+  }
+
+  RuntimeException wrapAlreadyClosedException(RuntimeException e) {
+    if (e instanceof NullPointerException) {
+      return alreadyClosed();
+    }
+    // TODO: maybe open a JDK issue to have a separate, more
+    // meaningful exception for unmapped segments:
+    if (e.getMessage() != null && e.getMessage().contains("closed")) {
+      return alreadyClosed();
+    }
+    return e;
+  }
+  
+  // return value only for compiler!
+  byte handlePositionalIOOBE(String action, long pos) throws IOException {
```

Comment:
The trick with byte return could be solved in a different way as well - by declaring a (dummy) return type of RuntimeException and replacing {{return handlePositionalIOOBE()}} with {{throw handlePositionalIOOBE(...)}} everywhere else. This has an additional benefit that code paths at invocation points see the code block as throwing an exception (so anything that follows is unreachable).

Reply:
Won't work here (the other code paths do this).
Here the problem is that it throws IOException or RuntimeRxception (IllegalArgumentEx extends RuntimeEx, EOFException extends checked IOException). There's no common Supertype. This is why option (2) was chosen here.

Reply:
The return also has same effect. Compiler understands that code is unreachable afterwards.

Please read my explanation as reply to Sokolov's question.

Reply:
I have. But I think the throws conveys it better :); the difference exists at call sites like this one:
{code}
    } catch (IndexOutOfBoundsException e) {
      handlePositionalIOOBE("seek", pos);
    }
{code}

The compiler would allow you to add code after the catch (or after handlePositionalOOBE). Whereas this would have the desired effect (also showing up-front that the method never returns normally):
{code}
    } catch (IndexOutOfBoundsException e) {
      throw handlePositionalIOOBE("seek", pos);
    }
{code}

Reply:
Ah now I understand... You declare it to return RuntimeException, but never return it. Instead throw exception directly, return type is just a fake.

Reply:
Precisely.

Reply:
Committed and pushed this change.